### PR TITLE
Feat(SCT-1056): Add endpoint for querying our new submissions

### PR DESCRIPTION
**Pull request by JohnFarrellDev**
## Link to JIRA ticket

[Ticket](https://hackney.atlassian.net/browse/SCT-1056)

### *What is the problem we're trying to solve*

We want a GET endpoint for our FormSubmissionController that works using query parameters. Based on query construct a request to mongo and return relevant CaseSubmissionResponse List.

### *What changes have we introduced*

Removed original GET /api/v1/submissions endpoint as this was just to return in-progress submissions.
Made a new GET /api/v1/submissions [FormSubmissionController.cs] endpoint that takes as query parameters FormId (string) and SubmissionStates (List<string>) [QueryCaseSubmission.cs]

Updated FormSubmissionUseCase, removed function for just getting in-progress submissions. Added function `ExecuteGetByQuery` which gets all the relevant forms based from MongoDB based on the query it receives.

#### _Checklist_

- [ ] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [ ] Added tests to cover all new production code
- [ ] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### *Follow up actions after merging PR*

Need to also merge in changes on the frontend that call the old GET api/v1/submissions endpoint that wants all submissions in-progress. Must be merged simultaneously.

### CODE REVIEW

File: SocialCareCaseViewerApi.Tests/V1/UseCase/FormSubmissionUseCaseTests.cs [hunk 1]
```diff
@@ -694,6 +663,78 @@ public void ExecuteUpdateSubmissionToApprovedThrowsUpdateSubmissionExceptionIfAp
...
+            var filter = builder.Empty;
+            filter &= Builders<CaseSubmission>.Filter.Eq(s => s.FormId, testFormId);
+
+            _mockMongoGateway.Setup(m =>
+                m.LoadRecordsByFilter(It.IsAny<string>(), It.IsAny<FilterDefinition<CaseSubmission>>()));
+
+            _formSubmissionsUseCase.ExecuteGetByQuery(request, filter);
+
+            _mockMongoGateway.Verify(x =>
+                x.LoadRecordsByFilter(MongoConnectionStrings.Map[Collection.ResidentCaseSubmissions], It.IsAny<FilterDefinition<CaseSubmission>>()), Times.Once);
```
Comment (JohnFarrellDev) [inline SocialCareCaseViewerApi.Tests/V1/UseCase/FormSubmissionUseCaseTests.cs:694]: Is there a way to improve this test to confirm the function was called with something equivalent to the filter object we constructed?

File: SocialCareCaseViewerApi.Tests/V1/UseCase/FormSubmissionUseCaseTests.cs [hunk 2]
```diff
@@ -694,6 +663,78 @@ public void ExecuteUpdateSubmissionToApprovedThrowsUpdateSubmissionExceptionIfAp
...
+            var filter = builder.Empty;
+            filter &= Builders<CaseSubmission>.Filter.In(s => s.SubmissionState, submissionStates);
+
+            _mockMongoGateway.Setup(m =>
+                m.LoadRecordsByFilter(It.IsAny<string>(), It.IsAny<FilterDefinition<CaseSubmission>>()));
+
+            _formSubmissionsUseCase.ExecuteGetByQuery(request, filter);
+
+            _mockMongoGateway.Verify(x =>
+                x.LoadRecordsByFilter(MongoConnectionStrings.Map[Collection.ResidentCaseSubmissions], It.IsAny<FilterDefinition<CaseSubmission>>()), Times.Once);
```
Comment (JohnFarrellDev) [inline SocialCareCaseViewerApi.Tests/V1/UseCase/FormSubmissionUseCaseTests.cs:714]: Is there a way to improve this test to confirm the function was called with something equivalent to the filter object we constructed?

File: SocialCareCaseViewerApi.Tests/V1/UseCase/FormSubmissionUseCaseTests.cs [hunk 3]
```diff
@@ -694,6 +663,78 @@ public void ExecuteUpdateSubmissionToApprovedThrowsUpdateSubmissionExceptionIfAp
...
+            filter &= Builders<CaseSubmission>.Filter.Eq(s => s.FormId, testFormId);
+            filter &= Builders<CaseSubmission>.Filter.In(s => s.SubmissionState, submissionStates);
+
+            _mockMongoGateway.Setup(m =>
+                m.LoadRecordsByFilter(It.IsAny<string>(), It.IsAny<FilterDefinition<CaseSubmission>>()));
+
+            _formSubmissionsUseCase.ExecuteGetByQuery(request, filter);
+
+            _mockMongoGateway.Verify(x =>
+                x.LoadRecordsByFilter(MongoConnectionStrings.Map[Collection.ResidentCaseSubmissions], It.IsAny<FilterDefinition<CaseSubmission>>()), Times.Once);
```
Comment (JohnFarrellDev) [inline SocialCareCaseViewerApi.Tests/V1/UseCase/FormSubmissionUseCaseTests.cs:736]: Is there a way to improve this test to confirm the function was called with something equivalent to the filter object we constructed?

File: SocialCareCaseViewerApi.Tests/V1/UseCase/FormSubmissionUseCaseTests.cs (hunk 1 above)
Comment (LBHTKarki) [reply SocialCareCaseViewerApi.Tests/V1/UseCase/FormSubmissionUseCaseTests.cs:694]: @JohnFarrellDev  Found some details about how to convert the filters to JSON format here https://jones.bz/testing-mongodb-queries/

So using the helper method: 
```
public static class FilterDefinitionExtensions
    {
        public static string RenderToJson<TDocument>(this FilterDefinition<TDocument> filter)
        {
            var serializerRegistry = BsonSerializer.SerializerRegistry;
            var documentSerializer = serializerRegistry.GetSerializer<TDocument>();
            return filter.Render(documentSerializer, serializerRegistry).ToJson();
        }
    }
```
it's possible to setup the gateway and verify the response based on the actual filter value:

 ```
...
var expectedJsonFilter = filter.RenderToJson();

  _mockMongoGateway.Setup(m =>
      m.LoadRecordsByFilter(It.IsAny<string>(), It.Is<FilterDefinition<CaseSubmission>>(filter => filter.RenderToJson().Equals(expectedJsonFilter))));

  _formSubmissionsUseCase.ExecuteGetByQuery(request, filter);

  _mockMongoGateway.Verify(x =>
      x.LoadRecordsByFilter(MongoConnectionStrings.Map[Collection.ResidentCaseSubmissions],
      It.Is<FilterDefinition<CaseSubmission>>(filter => filter.RenderToJson().Equals(expectedJsonFilter))), Times.Once);
```
That one option I suppose.

File: SocialCareCaseViewerApi/V1/UseCase/FormSubmissionsUseCase.cs [hunk 4]
```diff
@@ -56,15 +65,41 @@ public FormSubmissionsUseCase(IDatabaseGateway databaseGateway, IMongoGateway mo
             return foundSubmission?.ToDomain().ToResponse();
         }
 
-        public List<CaseSubmissionResponse> ExecuteListBySubmissionStatus(SubmissionState state)
+        public IEnumerable<CaseSubmissionResponse>? ExecuteGetByQuery(QueryCaseSubmissionsRequest request, FilterDefinition<CaseSubmission>? filter2 = null)
```
Comment (LBHTKarki) [inline SocialCareCaseViewerApi/V1/UseCase/FormSubmissionsUseCase.cs:68]: Is filter2 perhaps a left over from some earlier implementation? Looks like it's not used at the moment

File: SocialCareCaseViewerApi/V1/UseCase/FormSubmissionsUseCase.cs [hunk 5]
```diff
@@ -56,15 +65,41 @@ public FormSubmissionsUseCase(IDatabaseGateway databaseGateway, IMongoGateway mo
...
+            {
+                throw new QueryCaseSubmissionsException("Provide at minimum one query parameter");
+            }
+
+            var builder = Builders<CaseSubmission>.Filter;
+            var filter = builder.Empty;
+
+            if (request.FormId != null)
+            {
+                filter &= Builders<CaseSubmission>.Filter.Eq(s => s.FormId, request.FormId);
```
Comment (LBHTKarki) [inline SocialCareCaseViewerApi/V1/UseCase/FormSubmissionsUseCase.cs:80]: That's a neat way of building the filter and makes it easy to add more in the future 👍

**Review by LBHTKarki — COMMENTED** [review]
I think we can iterate on the test setup for the filters, but this is great for the intial search end point and will allow proper testing with reports

File: SocialCareCaseViewerApi.Tests/V1/UseCase/FormSubmissionUseCaseTests.cs (hunk 1 above)
Comment (JohnFarrellDev) [reply SocialCareCaseViewerApi.Tests/V1/UseCase/FormSubmissionUseCaseTests.cs:697]: This worked really well, thanks!

File: SocialCareCaseViewerApi/V1/UseCase/FormSubmissionsUseCase.cs [hunk 6]
```diff
@@ -56,15 +66,48 @@ public FormSubmissionsUseCase(IDatabaseGateway databaseGateway, IMongoGateway mo
             return foundSubmission?.ToDomain().ToResponse();
         }
 
-        public List<CaseSubmissionResponse> ExecuteListBySubmissionStatus(SubmissionState state)
+        public static string RenderToJson<TDocument>(FilterDefinition<TDocument> filter)
```
Comment (LBHTKarki) [inline SocialCareCaseViewerApi/V1/UseCase/FormSubmissionsUseCase.cs:69]: This was left here by mistake it seems

**Review by Renny4Real — APPROVED** [review]

**Pull request merged**